plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,239 @@
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <kvm/config.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <kvm-arm/smp_defs.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static uint64_t bsp_cpu_id;
+static unsigned int smp_numcores;
+void _lcpu_start(struct arm64_cpu *cpu);
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, cell;
+	unsigned int cpu_idx;
+	int subnode;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
+	/* Init the secondary cpus vector */
+	for (cpu_idx = 1; cpu_idx < CONFIG_UKPLAT_LCPU_MAXCOUNT; cpu_idx++) {
+		cpus[cpu_idx].id = (uint64_t) -1;
+		cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+	}
+	smp_numcores = 0;
+
+	/* Search for assigned VM cpus in DTB */
+	fdt_cpu = fdt_path_offset(_libkvmplat_cfg.dtb, "/cpus");
+	if (fdt_cpu < 0) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -1;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = 2;
+		uk_pr_warn("Using default value of 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
+		uk_pr_err("Invalid adress-cells size!\n");
+		return -1;
+	}
+
+	/* Search all the additional cpu nodes in DTB */
+	cpu_idx = 0;
+	fdt_for_each_subnode(subnode, _libkvmplat_cfg.dtb, fdt_cpu) {
+		int prop_len = 0;
+		uint64_t cpu_id;
+
+		cpu_prop = fdt_get_property(_libkvmplat_cfg.dtb, subnode,
```

Comment:
According to the spec, `disabled` means " quiescent state". Also, a disabled cpu can't be affected by running CPUs "except by an explicit method for enabling or re-enabling the quiescent CPU (see the enable-method property)".
Therefore, from my understanding, a `disabled` CPU isn't a CPU that is not meant to be used, but a CPU that is currently out of the game and should be "woke up" using the method specificed by the `enable-method` field.

Now, since `smp_init()` should be only run at boot time, I think all the cpus should have `state: disabled` at this point (in case this property is present), excepting the bootstrap CPU. Therefore, I think we could skip checking the `state` property and check directly for the `enable-method` one, which is anyway mandatory for the `ukplat_lcpu_wakeup()` implementation. Basically, this would mean that we can keep this part of the code unchaged.